messages.go

```diff
@@ -182,3 +182,157 @@ func NewMsg(msgType string, parent ComposedMsg) (ComposedMsg, error) {
 
 	return msg, nil
 }
+
+// Publish creates a new ComposedMsg and sends it back to the return identities over the
+// IOPub channel
+func (receipt *msgReceipt) Publish(msgType string, content interface{}) error {
+	msg, err := NewMsg(msgType, receipt.Msg)
+
+	if err != nil {
+		return err
+	}
+
+	msg.Content = content
+	return receipt.SendResponse(receipt.Sockets.IOPubSocket, msg)
+}
+
+// Reply creates a new ComposedMsg and sends it back to the return identities over the
+// Shell channel
+func (receipt *msgReceipt) Reply(msgType string, content interface{}) error {
+	msg, err := NewMsg(msgType, receipt.Msg)
+
+	if err != nil {
+		return err
+	}
+
+	msg.Content = content
+	return receipt.SendResponse(receipt.Sockets.ShellSocket, msg)
+}
+
+// MIMEDataBundle holds data that can be presented in multiple formats. The keys are MIME types
+// and the values are the data formatted with respect to it's MIME type. All bundle should contain
+// at least a "text/plain" representation with a string value.
+type MIMEDataBundle map[string]interface{}
+
+// NewTextMIMEDataBundle creates a MIMEDataBundle that only contains a text representation described
+// the the parameter 'value'
+func NewTextMIMEDataBundle(value string) MIMEDataBundle {
+	return MIMEDataBundle{
+		"text/plain": value,
+	}
+}
+
+// KernelStatus holds a kernel execution state, for status broadcast messages.
+type KernelStatus struct {
+	ExecutionState string `json:"execution_state"`
+}
+
+// PublishKernelStarting publishes a status message notifying front-ends that the kernel is
+// starting up.
+func (receipt *msgReceipt) PublishKernelStarting() {
+	receipt.Publish("status",
+		KernelStatus{
+			ExecutionState: "starting",
+		},
+	)
+}
+
+// PublishKernelBusy publishes a status message notifying front-ends that the kernel is
+// doing work.
+func (receipt *msgReceipt) PublishKernelBusy() {
+	receipt.Publish("status",
+		KernelStatus{
+			ExecutionState: "busy",
+		},
+	)
+}
+
+// PublishKernelIdle publishes a status message notifying front-ends that the kernel is
+// free.
+func (receipt *msgReceipt) PublishKernelIdle() {
+	receipt.Publish("status",
+		KernelStatus{
+			ExecutionState: "idle",
+		},
+	)
+}
+
+// ExecuteInput holds the source code being executed and the execution counter value
+// associated with source being run.
+type ExecuteInput struct {
+	ExecCount int    `json:"execution_count"`
+	Code      string `json:"code"`
+}
+
+// PublishExecutionInput publishes a status message notifying front-ends of what code is
+// currently being executed.
+func (receipt *msgReceipt) PublishExecutionInput(execCount int, code string) {
+	receipt.Publish("execute_input",
+		ExecuteInput{
+			ExecCount: execCount,
+			Code:      code,
+		},
+	)
+}
+
+// ExecuteResult holds the output to the 'ExecCount'th code execution.
+type ExecuteResult struct {
```

Comment:
I'm not convinced that we need a type for each published result, like `ExecuteResult`, etc.  Maybe I'm wrong here, but I find myself jumping around trying to figure out what these things are.  Would it be more readable to just pass a map from these methods or define the struct in place when sending it to publish?